File: 图/Floyd.java
import java.util.Arrays;
import java.util.Scanner;

/** Floyd算法 --最短路径算法
 * Created by han on 2017/1/2.
 */
public class Floyd {


    public    final  int MAXVEX=100;
    public    final  int MAXSIZE=5;
    public    final  int MAX=100;
    public    final  int INFINITY=65535;

    //  邻接矩阵

    class MGraph{
        String[] vexs=new String[MAXVEX];
        int[][] arc = new int[MAXVEX][MAXVEX];
        int numVertexes,numEdges;

        @Override
        public String toString() {
            return "MGraph{" +
                    "vexs=" + Arrays.toString(vexs) +
                    ", arc=" + Arrays.toString(arc) +
                    ", numVertexes=" + numVertexes +
                    ", numEdges=" + numEdges +
                    '}';
        }
    }


    class  Queue
    {
        Object[] data = new Object[MAXSIZE];
        int front;
        int rear;
    }

    boolean InitQueue(Queue Q){
        Q.front=0;
        Q.rear=0;
        return  true;
    }

    boolean EnQueue(Queue Q,Object e){
        if((Q.rear+1)%MAXSIZE ==Q.front)
            return  false;
        Q.data[Q.rear]=e;
        Q.rear=(Q.rear+1)%MAXSIZE;

        return  true;
    }


    Object DeQueue(Queue Q){
        Object e;
        if(Q.front ==Q.rear)
            return  false;
        e=Q.data[Q.front];
        Q.front=(Q.front+1)%MAXSIZE;
        return  e;
    }


    boolean QueueEmpty(Queue Q){

        if(Q.front==Q.rear)
            return true;
        else
            return false;
    }


    void CreateMGrapht(MGraph G){

        int i,j,k,w;

        System.out.println("输入顶点数和边数");
        Scanner sc = new Scanner(System.in);
        G.numVertexes = sc.nextInt();
        G.numEdges = sc.nextInt();
        System.out.println("输入顶点信息");
        for (i=0;i<G.numVertexes;i++)
            G.vexs[i] = sc.next();
        for (i=0;i<G.numVertexes;i++)
            for (j=0;j<G.numVertexes;j++)
                G.arc[i][j]=INFINITY;

        for (k=0;k<G.numEdges;k++){
            System.out.println("输入边(vi,vj)上的下标i,上标j和权w");
            i = sc.nextInt();
            j = sc.nextInt();
            w=sc.nextInt();
            G.arc[i][j]=w;
            G.arc[j][i]=G.arc[i][j];
            System.out.println(G.arc[i][j]);
        }
    }

     void  ShortestPath_Floyd(MGraph G,int[][] pathmatirx,int[][] ShortPathTable){
         int v,w,k;
         for (v=0;v<G.numVertexes;++v){
           for (w=0;w<G.numVertexes;++w) {
               ShortPathTable[v][w] = G.arc[v][w];
               pathmatirx[v][w]=w;
           }
         }

         for (k=0;k<G.numVertexes;++k){
             for (v=0;v<G.numVertexes;++v){
                 for (w=0;w<G.numVertexes;++w){
                     if(ShortPathTable[v][w] >ShortPathTable[v][k]+ShortPathTable[k][w]){
                          ShortPathTable[v][w]=ShortPathTable[v][k]+ShortPathTable[k][w];
                          pathmatirx[v][w]=pathmatirx[v][k];
                     }
                 }
             }
         }


         //求最短路径的显示
         for (v=0;v<G.numVertexes;++v){
             for (w=v+1;w<G.numVertexes;w++) {
                 System.out.println();
                 System.out.print("V:" + v + " V:" + w + " Weight[" + v + "]" + "[" + w + "]：" + ShortPathTable[v][w]);
                 k=pathmatirx[v][w];
                 System.out.print(" path:"+v);
                 while (k!=w){
                     System.out.print("->"+k);
                     k=pathmatirx[k][w];
                 }
                 System.out.print("->"+w);
             }
             System.out.print("\n");
         }

     }

    public static void main(String[] args) {
        Floyd floyd = new Floyd();
        MGraph G = new Floyd().new MGraph();
        floyd.CreateMGrapht(G);
        floyd.ShortestPath_Floyd(G, new int[floyd.MAXVEX][floyd.MAXVEX], new int[floyd.MAXVEX][floyd.MAXVEX]);
    }
}

File: 大话数据结构原书C版以及Java版翻译/图/Floyd.c
#include <stdio.h>
#define OK 1
#define ERROR 0
#define TRUE 1
#define FALSE 0
#define MAXSIZE 100
#define MAXEDGE 20
#define MAXVEX 100  /*最大顶点数 应由用户定义*/
#define INFINITY 65535 /*用65535表示OO*/
typedef char VertexType; /*顶点类型应由用户定义*/
typedef int  EdgeType; /*边上的权值类型应由用户定义*/
typedef int QElemType;
#define    MAX 100
typedef  int  Boolean;/*Boolean是布尔类型 其值是True或False*/
typedef int Status;
typedef struct {
	VertexType vexs[MAXVEX];/*顶点表*/
	EdgeType arc[MAXVEX][MAXVEX];/*临接矩阵 可看成表*/
	int numVertexes,numEdges;/*图中当前的顶点数和边数*/
} MGraph;


/*对边集数组*/
typedef struct {
	int begin;
	int end;
	int weight;
} Edge;


typedef int Pathmatirx[MAXVEX][MAXVEX];
typedef int ShortPathTable[MAXVEX][MAXVEX];
/*Floyd算法 求网图G中各顶点v到其余顶点w最短路径p[v][w]及带权长度D[v][w]*/
void ShortestPath_Floyd(MGraph G,Pathmatirx *P,ShortPathTable *D) {
	int v,w,k;
	for(v=0; v<G.numVertexes; ++v) { /*初始化D与P*/
		for(w=0; w<G.numVertexes; ++w) {
			(*D)[v][w]=G.arc[v][w]; /*D[v][w]值即为对应点间的权值*/
			(*P)[v][w]=w;    /*初始化P*/
		}
	}

	for(k=0; k<G.numVertexes; ++k) {
		for(v=0; v<G.numVertexes; ++v) {
			for(w=0; w<G.numVertexes; ++w) {
				if((*D)[v][w]>(*D)[v][k]+(*D)[k][w]) {
					/*如果经过下标为k顶点路径比原俩点路径更短*/
					/*将当前俩点间权值设为更小的一个*/
					(*D)[v][w]=(*D)[v][k]+(*D)[k][w];
					(*P)[v][w]=(*P)[v][k]; /*路径设置经过下标为k的顶点*/
				}
			}
		}
	}

	//求最短路径的显示
	for(v=0; v<G.numVertexes; ++v) {
		for(w=v+1; w<G.numVertexes; w++) {
			printf("v%d-v%d weight: %d",v,w,(*D)[v][w]);
			k=(*P)[v][w];         //获得第一个路径顶点下标
			printf("path : %d",v); //打印源点
			while(k!=w) { //如果路径顶点下标不是终点
				printf("-> %d",k);//打印路径顶点
				k=(*P)[k][w]; //获得下一个路径顶点下标
			}
			printf("-> %d\n",w);//打印终点
		}
		printf("\n");
	}
}
